modules/core/src/main/scala/org/tesselation/infrastructure/trust/TrustDaemon.scala

```diff
@@ -0,0 +1,53 @@
+package org.tesselation.infrastructure.trust
+
+import cats.Parallel
+import cats.effect.std.Random
+import cats.effect.{Async, Spawn, Temporal}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.parallel._
+
+import org.tesselation.config.types.TrustDaemonConfig
+import org.tesselation.domain.Daemon
+import org.tesselation.domain.cluster.storage.ClusterStorage
+import org.tesselation.domain.trust.storage.TrustStorage
+import org.tesselation.http.p2p.P2PClient
+import org.tesselation.keytool.security.SecurityProvider
+import org.tesselation.kryo.KryoSerializer
+
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait TrustDaemon[F[_]] extends Daemon[F]
+
+object TrustDaemon {
+
+  def make[F[_]: Async: SecurityProvider: KryoSerializer: Random: Parallel](
+    cfg: TrustDaemonConfig,
+    p2pClient: P2PClient[F],
+    clusterStorage: ClusterStorage[F],
+    trustStorage: TrustStorage[F]
+  ): TrustDaemon[F] = new TrustDaemon[F] {
+
+    private val logger = Slf4jLogger.getLogger[F]
+
+    def start: F[Unit] =
+      for {
+        _ <- Spawn[F].start(pollTrust.foreverM).void
+      } yield ()
+
+    // Temporary stub, move to separate program.
+    private def calculateTrust(): F[Unit] =
+      Async[F].pure(())
+
+    private def pollTrust: F[Unit] =
+      for {
+        _ <- Temporal[F].sleep(cfg.interval)
+        peers <- clusterStorage.getPeers
+        _ <- peers.toList.parTraverse { peer =>
+          p2pClient.trust.getPublicTrust.run(peer).map(pt => trustStorage.updatePeerPublicTrustInfo(peer.id, pt))
+        }
```

Comment:
> To clarify: what we need here is a chain of signatures

Chain of signatures is a thing required at a totally different layer and for different purpose. Here, it’s not needed at all, and the result will be the same, it’s still secured by the algorithms I’ve described above (replay attack not possible)

Reply:
@marcinwadon 

Re: signatures in gossip. Reviewed the code and I see it now, so that rules out some of the attacks I mentioned. 

>no, because we track each message and handle it exactly once only. and there is always a way to re-check and track it in the handler, when more security is needed.

Regarding this -- can you provide more detail exactly on where this is implemented (I am unfamiliar with this section of the code -- and I don't see it at a glance. I just see hash and signature validators, is this on a different branch?) Also -- can you verify if the hashes are based purely on the value of the data? Because if so, trust updates can actually have the same data multiple times, e.g. saying node_a 50%, node_a 40%, node_a 50% as three separate updates would have a duplicate for message 1 and 3. You mention the counters later on -- that's what I and Wyatt mean by a dependency hash chain. (Irrelevant whether it's a counter or a hash reference -- hash references are easier for verifying against storage)

I'm not arguing against the scalability problems here. I'd prefer to use a gossip based solution as long as we have an appropriate guarantee of defense against replays -- but the model as implemented is not complete, and should be prioritized. I imagine the broadcast is quite terrible for messages that need to be repeated much more often -- but trust updates can be incredibly slow and still provide security.

A partial solution for dealing with trust updates violating the state is simply to ban any node that signs an invalid chain -- that avoids dealing with any consensus at all. If there are two signatures of the same state it's relatively easy to prove that there's something wrong -- no agreement is necessary, which means this doesn't require a full implementation that connects to the rest of the code. Any update can simply check the previous update. 

Again, right now I'm still reviewing the replay attack protection to try to verify what you said above. I'll discuss prioritization of this with Wyatt. I was originally trying to just get feature parity so I can work on improvements to the core model. If the long delay on trust updates is insufficient, I can work on the final implementation of gossiped trust -- it doesn't require any other components if I implement 'lastHash' update essentially, but it's more work than feature parity.

Reply:
@ryle-goehausen 
>can you provide more detail exactly on where this is implemented (I am unfamiliar with this section of the code -- and I don't see it at a glance. I just see hash and signature validators, is this on a different branch?)

We track `seen` rumors in `RumorStorage`. If we receive a rumor we've already seen, we're just going to ignore it, the implementation of this filtering can be found in `setRumorsSeenAndGetUnseen` method.

It's worth mentioning that there's a retention time on hashes of seen rumors, currently set to 2 minutes - so currently in practice after 2 minutes it will be possible to replay the message and the receiver will process it as it was never seen before. We can solve this issue however with retention based on message counters instead of a time @marcinwadon is working on adding counters to the message.

Reply:
I'd like to add that having a chain of messages would require us to keep all the historical data. To avoid this, the Transaction and the Fiber forms a chain using counters as well (ordinal). That way we can quickly verify whether the transaction or fiber is correct in order, by comparing the ordinal number.

Reply:
@ryle-goehausen see: #56

Reply:
Hey guys, talked with Wyatt -- I'm going to change this PR to use the gossip since you are handling the sequence numbers (the replay defense.)

Re: retention time on hashes -- I'm going to just add a randomization factor to the trust message to get around the issue of a legitimate user changing a score back and forth -- so it doesn't accidentally generate the same legitimate hash twice and ignore the message.

>I'd like to add that having a chain of messages would require us to keep all the historical data. To avoid this, the Transaction and the Fiber forms a chain using counters as well (ordinal). 

For the other gossip messages with arbitrary complexity yes -- for a trust update we're already storing the latest version of it (so storing a hash is easy,) and it only ever compares one value. I fully agree though that this is the equivalent of a sequence number on gossip messages (which you have implemented.) I was thinking I was going to need to implement this logic myself -- which is why hash is easy to do as just 'lastTrustUpdate'. Sequence number works exactly the same in this case, since you're only comparing one value.

Libp2p has the exact same mechanism -- see here: https://github.com/libp2p/specs/pull/217/files so this has already been reviewed elsewhere and is valid.